### DISCUSSION TITLE
Want to support trayMenu

### DISCUSSION THREAD

**Discussion by daodao97** [wailsapp/wails#1438] [Q&A] [answered]
Are there any plans to support system trayMenu?

I find [xbar](https://github.com/matryer/xbar) can set trayMenu, it's wails version is `v2.0.0-alpha.72`.

But the last wails version `v2.0.0-beta.37` not support it.

![](https://github.com/matryer/xbar/raw/main/xbarapp.com/public/img/xbar-menu-preview.png)

**Comment by stffabi** [comment]
Tray support is planned for a post v2 release, IIRC for v2.1

**Reply by daodao97** [reply]
look forward for it, what is the time will release v2.1?

**Reply by stffabi** [reply]
AFAIK there's no planned time frame for that.

**Comment by ironpark** [comment, marked as answer]
@daodao97 @stffabi 

Currently, wails does not support the tray menu, and both the systray library and wails use the main thread, so it is difficult to use them together.

so, I made a small program that uses the systray library. I also created a client library that can be controlled via IPC.

```
[Remote Tray] -- Local IPC -- [Wails App]
```

The library automatically installs the remote control tray application in the temporary folder, and after running it, set up IPC communication.
``` go
func RunTray(){
    tray, err := remotray.Run("<unix-domain-socket | pipe name for IPC>",
        remotray.WithTitle("TITLE"),)
    if err != nil {
        panic(err)
    }
    item, _ := tray.AddMenuItem("item title.1", "Tooltop")
    item.OnClick(func(item MenuItem) {
        fmt.Println("onclick", item)
    })
    item, _ = tray.AddMenuItem("quit", "Tooltop")
    item.OnClick(func(item MenuItem) {
        tray.Quit()
    })
}
```
finally, I was able to control the tray menu while using the wails library.

https://github.com/Ironpark/remotray
The library is currently in POC state and has only been tested on MacOS.

**Reply by daodao97** [reply]
Great, I will try it in my app [EasyClash](https://github.com/daodao97/EasyClash)

In addition, I would like to know the difference with [systray](https://github.com/getlantern/systray)

**Reply by ironpark** [reply]
https://github.com/Ironpark/remotray/blob/main/internal/remotray/tray.go
The library uses systray internally, main difference is that the process controlling traymenu runs as a separate process. also, since protocols for all functions of the systray library have not been implemented, only some functions are currently available.

> P.S reason I used the fyne.io/systray fork rather than github.com/getlantern/systray is because it is more actively maintained.

**Reply by leaanthony** [reply]
Tray menu is on the roadmap and already works for Mac on a branch. When we get v2 out the door let's focus on getting that done 👍
